src/main/java/org/quiltmc/loader/impl/config/tree/Trie.java

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.impl.config.tree;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.ConcurrentModificationException;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.jetbrains.annotations.NotNull;
+import org.quiltmc.loader.api.config.TrackedValue;
+import org.quiltmc.loader.api.config.values.ValueKey;
+import org.quiltmc.loader.api.config.values.ValueTreeNode;
+import org.quiltmc.loader.impl.config.builders.SectionBuilderImpl;
+import org.quiltmc.loader.impl.config.values.ValueKeyImpl;
+
+public final class Trie {
+	private final Node root = new Node(null, null, null);
+
+	private int modCount;
+
+	public Iterable<ValueTreeNode> leaves() {
+		return LeafItr::new;
+	}
+
+	public Iterable<ValueTreeNode> nodes() {
+		return new Iterable<ValueTreeNode>() {
+			@NotNull
+			@Override
+			public Iterator<ValueTreeNode> iterator() {
+				return new Iterator<ValueTreeNode>() {
+					private final Iterator<Node> itr = Trie.this.root.iterator();
+
+					@Override
+					public boolean hasNext() {
+						return this.itr.hasNext();
+					}
+
+					@Override
+					public ValueTreeNode next() {
+						return this.itr.next().getValue();
+					}
+				};
+			}
+		};
+	}
+
+	public ValueTreeNode get(String key0, String... keys) {
+		return this.root.getOrCreateNChild(key0, keys).value;
+	}
+
+	public ValueTreeNode put(ValueTreeNode value, String key0, String... keys) {
+		int modifiedCount = this.modCount;
+
+		Node node = this.root.getOrCreateNChild(key0, keys);
+		ValueTreeNode oldValue = node.getValue();
+		node.setValue(value);
+
+		// Only increment the number of modifications if a new node wasn't created by the call to getOrCreateNChild
+		if (modifiedCount == this.modCount) {
+			++this.modCount;
+		}
+
+		return oldValue;
+	}
+
+	public ValueTreeNode put(ValueKey key, ValueTreeNode value) {
```

Comment:
Fair enough, this method might as well take an `Iterable` too. The argument order is reversed on the other method so that it can use varargs for the key. That method isn't even used though, so I guess I might as well remove it.